File: Compiler/src/compiler/operators/CheckClass.java
package compiler.operators;

import compiler.Operator;
import compiler.values.ClassValue;
import dictionaries.ClassesDictionary;
import util.CompilationResult;
import util.DataType;
import util.JenaUtil;
import util.Pair;

import java.util.ArrayList;
import java.util.List;

/**
 * Оператор проверки класса объекта
 */
public class CheckClass extends BaseOperator {

    /**
     * Имя предиката, используемое при компиляции
     */
    private static final String CLASS_PRED_NAME = "type";

    /**
     * Конструктор
     * @param args Аргументы
     */
    public CheckClass(List<Operator> args) {
        super(args);
    }

    @Override
    public List<List<DataType>> argsDataTypes() {
        List<List<DataType>> result = new ArrayList<>();

        result.add(List.of(DataType.OBJECT, DataType.CLASS));

        return result;
    }

    @Override
    public DataType resultDataType() {
        return DataType.BOOLEAN;
    }

    @Override
    public CompilationResult compile() {
        // Объявляем переменные
        String value = "";
        String ruleHead = "";
        String completedRules = "";

        // Получаем аргументы
        Operator arg0 = arg(0);
        Operator arg1 = arg(1);

        // Компилируем аргументы
        CompilationResult compiledArg0 = arg0.compile();
        CompilationResult compiledArg1 = arg1.compile();

        // Если класс можно вычислить (вычисляемый класс можно получить только указав его имя т.е. через ClassValue)
        if(arg1 instanceof ClassValue &&
                ClassesDictionary.isComputable(((ClassValue) arg1).value())) {
            // Получаем выражение для вычисления
            Pair<String, Operator> calculation = ClassesDictionary.howToCalculate(((ClassValue) arg1).value());
            String varName = calculation.first();
            Operator expression = calculation.second();

            CompilationResult compiledCalculation = expression.compile(); // Выражение для вычисления

            // Собираем правило
            ruleHead = compiledArg0.ruleHead() + compiledArg1.ruleHead(); // Собираем части первого и второго аргументов
            ruleHead += JenaUtil.genBindPrim(compiledArg0.value(), varName); // Инициализируем переменную
            ruleHead += compiledCalculation.ruleHead(); // Добавляем результат компиляции вычисления

            // Передаем завершенные правила дальше
            completedRules = compiledArg0.completedRules() +
                    compiledArg1.completedRules() +
                    compiledCalculation.completedRules();
        }
        else {
            // Собираем правило
            ruleHead = compiledArg0.ruleHead() + compiledArg1.ruleHead(); // Собираем части первого и второго аргументов

            // Добавляем проверку класса
            ruleHead += JenaUtil.genTriple(
                    compiledArg0.value(),
                    JenaUtil.genLink(JenaUtil.RDF_PREF, CLASS_PRED_NAME),
                    compiledArg1.value());

            // Передаем завершенные правила дальше
            completedRules = compiledArg0.completedRules() + compiledArg1.completedRules();
        }

        usedObjects = List.of(compiledArg0.value());

        return new CompilationResult(value, ruleHead, completedRules);
    }
}

File: Compiler/src/compiler/operators/CheckClass.kt
package compiler.operators

import compiler.Operator
import compiler.literals.ClassLiteral
import compiler.util.CompilationResult
import dictionaries.ClassesDictionary.calcExpr
import dictionaries.ClassesDictionary.isCalculable
import models.ClassModel
import util.DataType
import util.JenaUtil
import util.JenaUtil.CLASS_PREDICATE_NAME
import util.JenaUtil.genBindPrim
import util.JenaUtil.genLink
import util.JenaUtil.genNoValuePrim
import util.JenaUtil.genTriple
import util.JenaUtil.genVar

/**
 * Оператор проверки класса объекта
 */
class CheckClass(args: List<Operator>) : BaseOperator(args) {

    /**
     * Является ли оператор негативным (т.е. нужно ли отрицание при компиляции)
     */
    internal var isNegative = false

    override val argsDataTypes
        get() = listOf(listOf(DataType.Object, DataType.Class))

    override val resultDataType
        get() = DataType.Boolean

    override fun compile(): CompilationResult {
        // Объявляем переменные
        val bodies = ArrayList<String>()
        var completedRules = ""

        // Получаем аргументы
        val arg0 = arg(0)
        val arg1 = arg(1)

        // Компилируем аргументы
        val compiledArg0 = arg0.compile()
        val compiledArg1 = arg1.compile()

        // Если класс можно вычислить (вычисляемый класс можно получить только указав его имя т.е. через ClassValue)
        if (isCalculable((arg1 as ClassLiteral).value)) {
            // Получаем выражение для вычисления
            val calculation = calcExpr(arg1.value)

            // Проверяем корректность словаря
            requireNotNull(calculation) { "Для класса ${arg1.value} в словаре нет выражения" }

            val varName = ClassModel.CALC_EXPR_VAR_NAME
            var expression = calculation

            // Если негативная форма - добавляем отрицание
            if (isNegative) {
                expression = LogicalNot(listOf(expression))
            }

            // Компилируем выражение для вычисления
            val compiledCalculation = expression.semantic().compile()

            // Передаем завершенные правила дальше
            completedRules += compiledArg0.rules +
                    compiledArg1.rules +
                    compiledCalculation.rules

            // Для всех результатов компиляции
            compiledArg0.bodies.forEach { body0 ->
                compiledArg1.bodies.forEach { body1 ->
                    compiledCalculation.bodies.forEach { calculationBody ->
                        // Собираем правило
                        var body = body0 + body1 // Собираем части первого и второго аргументов
                        body += genBindPrim(compiledArg0.value, genVar(varName)) // Инициализируем переменную
                        body += calculationBody // Добавляем результат компиляции вычисления

                        // Добавляем в массив
                        bodies.add(body)
                    }
                }
            }
        } else {
            // Передаем завершенные правила дальше
            completedRules += compiledArg0.rules +
                    compiledArg1.rules

            // Для всех результатов компиляции
            compiledArg0.bodies.forEach { body0 ->
                compiledArg1.bodies.forEach { body1 ->
                    // Собираем правило
                    var body = body0 + body1 // Собираем части первого и второго аргументов

                    // Добавляем проверку класса
                    body += if (isNegative) {
                        // Если форма негативная - проверяем отсутствие класса
                        genNoValuePrim(
                            compiledArg0.value,
                            genLink(JenaUtil.RDF_PREF, CLASS_PREDICATE_NAME),
                            compiledArg1.value
                        )
                    } else {
                        // Проверяем наличие класса
                        genTriple(
                            compiledArg0.value,
                            genLink(JenaUtil.RDF_PREF, CLASS_PREDICATE_NAME),
                            compiledArg1.value
                        )
                    }

                    // Добавляем в массив
                    bodies.add(body)
                }
            }
        }

        return CompilationResult(bodies = bodies, rules = completedRules)
    }

    override fun clone(): Operator {
        val newArgs = ArrayList<Operator>()

        args.forEach { arg ->
            newArgs.add(arg.clone())
        }

        return CheckClass(newArgs)
    }

    override fun clone(newArgs: List<Operator>): Operator {
        return CheckClass(newArgs)
    }
}